[DEVOPS-398] Public testnet deployment

rvl opened the pull request:
This is for the new testnet-deployer host and a module for cardano-sl-faucet.

deepfire commented on deployments/testnet-deployer.nix:
```diff
@@ -0,0 +1,86 @@
+{ ... }:
```
Great that we now finally have the deployer separate!

deepfire commented on lib/faucet.html:
```diff
@@ -0,0 +1,182 @@
+<!doctype html>
```
Does this belong to a separate repository?

disassembler replied:
I agree, I think we should have a faucet-ui repo (or maybe a derivation on cardano-sl, since that's where the main faucet code is) or something instead of embedding html/vuejs code in iohk-ops.

rvl replied:
Will move it to `cardano-sl/faucet/frontend` and add a derivation there.

disassembler commented on deployments/report-server-env-testnet.nix:
```diff
@@ -0,0 +1,20 @@
+{ globals, ... }: with (import ./../lib.nix);
```
is this needed considering zendesk only works with mainnet?

rvl replied:
Good point -- it is unneeded. I have removed this and killed the obsolete nixops resources.

disassembler commented on deployments/security-groups.nix:
```diff
@@ -119,6 +120,15 @@ with import ../lib.nix;
+              description = "Access Cardano Faucet";
+              rules = [{
+                protocol = "tcp";
+                fromPort = 80; toPort = 80;
```
should be 443/ssl instead of allowing insecure 80 traffic

rvl replied:
OK I will add 443 and leave 80 to allow redirecting http -> https.

disassembler commented on lib.nix:
```diff
@@ -50,44 +50,19 @@ in lib // (rec {
-    rodneyKey
-    samKey
-  ];
+  inherit devOps csl-developers;
```
love this abstraction!

disassembler commented on modules/cardano-faucet.nix:
```diff
@@ -0,0 +1,55 @@
+  explorerURL = "http://cardano-explorer.cardano-testnet.iohkdev.io/";
+  faucetFrontend = pkgs.runCommand "faucet-frontend" {} ''
+    mkdir -p $out
+    sed -e 's|recaptchaSiteKey:.*|recaptchaSiteKey: "${recaptcha.siteKey}",|g' \
```
I don't like that we're reading the recaptcha key from a private static area, but then injecting it plain text into the store. Essentially, if the server is hacked under any user, this makes the key visible. Couldn't we deploy this as a deploy key and have the faucet take a path to the file containing the key so it doesn't end up in the nix store? Since we own the faucet code, I don't see why we wouldn't want to go this route.

rvl replied:
True. The recaptcha site key is public and injected into HTML. The recaptcha secret key is private and shouldn't be in the nix store. I already asked if faucet code could read the recaptcha secret key from a file but everything has to be asked for at least twice.

disassembler commented on modules/cardano-faucet.nix:
```diff
@@ -0,0 +1,55 @@
+
+  networking.firewall.allowedTCPPorts = [ 80 ];
+
+  services.nginx = {
```
lets set this up as SSL. `enableACME = true;` will get you a letsencrypt cert at a minimum

disassembler commented on modules/cardano-faucet.nix:
```diff
@@ -0,0 +1,55 @@
+    inherit recaptcha;
+  };
+
+  networking.firewall.allowedTCPPorts = [ 80 ];
```
I think we only want to expose 80 if we setup a forced ssl redirect (`forceSSL = true;`). 443 should be added here as well.

rvl replied:
OK - will add both 80 and 443 and use forced ssl redirect.

disassembler commented on modules/cardano-faucet.nix:
```diff
@@ -0,0 +1,55 @@
+        then config.global.dnsDomainname else "iohkdev.io";
+     in {
+      "cardano-faucet.${vhostDomainName}" = {
+        # TLS provided by cloudfront
```
even if we have SSL terminating at cloudfront, it would still be good to have SSL from cloudfront to our server. https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-https-cloudfront-to-custom-origin.html

rvl replied:
Yes good point. Anyway I think cloudfront doesn't make sense for faucet but might for explorer (where this comment was copied from). So I will delete the comment and add enableACME.

disassembler commented on modules/cardano-faucet.nix:
```diff
@@ -0,0 +1,55 @@
+        locations = {
+          "/" = {
+            root = faucetFrontend;
+            tryFiles = "$uri $uri/index.html @faucet";
```
what about error pages? What do we want the user to see if the backend isn't available, or takes to long to respond, etc...

rvl replied:
The standard horrible error pages from NGINX should be OK.

This is because the faucet will normally be called by XHR through a page like this https://testnet.iohkdev.io/goguen/faucet/ . This frontend should show some "unavailable" message when it gets 5xx responses.

disassembler commented on modules/cardano-faucet-module.nix:
```diff
@@ -0,0 +1,268 @@
+    source-wallet.generate-to = "${cfg.home}/generated-wallet.json";
+  } // (optionalAttrs (cfg.recaptcha != null) {
+    # fixme: recaptcha secret in file
+    recaptcha-secret = cfg.recaptcha.secretKey;
```
good point on public vs secret recaptcha keys, but this is where the secret is getting dumped to the nix store.

rvl replied:
Faucet has now been fixed to read the secret key from file so I can change this.

rvl commented:
Thanks for the reviews. New commits have been pushed to this PR. I believe that I have addressed all review comments. The scripts are also shellcheck clean.

rvl commented:
Rebased on latest develop.

rvl commented:
Rebased onto master, resolved merge conflicts, removed change to `cardano-sl-src.json`, added CORS to faucet NGINX.
